File: c/anharmonic/phonon3/reciprocal_to_normal.c
/* Copyright (C) 2015 Atsushi Togo */
/* All rights reserved. */

/* This file is part of phonopy. */

/* Redistribution and use in source and binary forms, with or without */
/* modification, are permitted provided that the following conditions */
/* are met: */

/* * Redistributions of source code must retain the above copyright */
/*   notice, this list of conditions and the following disclaimer. */

/* * Redistributions in binary form must reproduce the above copyright */
/*   notice, this list of conditions and the following disclaimer in */
/*   the documentation and/or other materials provided with the */
/*   distribution. */

/* * Neither the name of the phonopy project nor the names of its */
/*   contributors may be used to endorse or promote products derived */
/*   from this software without specific prior written permission. */

/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS */
/* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT */
/* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS */
/* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE */
/* COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, */
/* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, */
/* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; */
/* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER */
/* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT */
/* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN */
/* ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE */
/* POSSIBILITY OF SUCH DAMAGE. */

#include <stdlib.h>
#include <math.h>
#include <phonoc_utils.h>
#include <phonoc_const.h>
#include <phonoc_array.h>
#include <phonon3_h/reciprocal_to_normal.h>
#include <lapack_wrapper.h>

#ifdef MEASURE_R2N
#include <unistd.h>
#include <time.h>
#endif

static lapack_complex_double fc3_sum_in_reciprocal_to_normal
(const size_t bi0,
 const size_t bi1,
 const size_t bi2,
 const lapack_complex_double *eigvecs0,
 const lapack_complex_double *eigvecs1,
 const lapack_complex_double *eigvecs2,
 const lapack_complex_double *fc3_reciprocal,
 const double *masses,
 const size_t num_atom);

static double get_fc3_sum
(const size_t j,
 const size_t k,
 const size_t bi,
 const double *freqs0,
 const double *freqs1,
 const double *freqs2,
 const lapack_complex_double *eigvecs0,
 const lapack_complex_double *eigvecs1,
 const lapack_complex_double *eigvecs2,
 const lapack_complex_double *fc3_reciprocal,
 const double *masses,
 const size_t num_atom,
 const double cutoff_frequency);

void reciprocal_to_normal_squared
(double *fc3_normal_squared,
 PHPYCONST int (*g_pos)[4],
 const size_t num_g_pos,
 const lapack_complex_double *fc3_reciprocal,
 const double *freqs0,
 const double *freqs1,
 const double *freqs2,
 const lapack_complex_double *eigvecs0,
 const lapack_complex_double *eigvecs1,
 const lapack_complex_double *eigvecs2,
 const double *masses,
 const int *band_indices,
 const size_t num_band0,
 const size_t num_band,
 const double cutoff_frequency,
 const int openmp_at_bands)
{
  size_t i, num_atom;

#ifdef MEASURE_R2N
  double loopTotalCPUTime, loopTotalWallTime;
  time_t loopStartWallTime;
  clock_t loopStartCPUTime;
#endif

  num_atom = num_band / 3;

#ifdef MEASURE_R2N
  loopStartWallTime = time(NULL);
  loopStartCPUTime = clock();
#endif

#pragma omp parallel for if (openmp_at_bands)
  for (i = 0; i < num_g_pos; i++) {
    if (freqs0[band_indices[g_pos[i][0]]] > cutoff_frequency) {
      fc3_normal_squared[g_pos[i][3]] = get_fc3_sum(g_pos[i][1],
                                                    g_pos[i][2],
                                                    band_indices[g_pos[i][0]],
                                                    freqs0,
                                                    freqs1,
                                                    freqs2,
                                                    eigvecs0,
                                                    eigvecs1,
                                                    eigvecs2,
                                                    fc3_reciprocal,
                                                    masses,
                                                    num_atom,
                                                    cutoff_frequency);
    }
  }

#ifdef MEASURE_R2N
      loopTotalCPUTime = (double)(clock() - loopStartCPUTime) / CLOCKS_PER_SEC;
      loopTotalWallTime = difftime(time(NULL), loopStartWallTime);
      printf("  %1.3fs (%1.3fs CPU)\n", loopTotalWallTime, loopTotalCPUTime);
#endif

}

static double get_fc3_sum
(const size_t j,
 const size_t k,
 const size_t bi,
 const double *freqs0,
 const double *freqs1,
 const double *freqs2,
 const lapack_complex_double *eigvecs0,
 const lapack_complex_double *eigvecs1,
 const lapack_complex_double *eigvecs2,
 const lapack_complex_double *fc3_reciprocal,
 const double *masses,
 const size_t num_atom,
 const double cutoff_frequency)
{
  double fff, sum_real, sum_imag;
  lapack_complex_double fc3_sum;

  if (freqs1[j] > cutoff_frequency && freqs2[k] > cutoff_frequency) {
    fff = freqs0[bi] * freqs1[j] * freqs2[k];
    fc3_sum = fc3_sum_in_reciprocal_to_normal
      (bi, j, k,
       eigvecs0, eigvecs1, eigvecs2,
       fc3_reciprocal,
       masses,
       num_atom);
    sum_real = lapack_complex_double_real(fc3_sum);
    sum_imag = lapack_complex_double_imag(fc3_sum);
    return (sum_real * sum_real + sum_imag * sum_imag) / fff;
  } else {
    return 0;
  }
}

static lapack_complex_double fc3_sum_in_reciprocal_to_normal
(const size_t bi0,
 const size_t bi1,
 const size_t bi2,
 const lapack_complex_double *eigvecs0,
 const lapack_complex_double *eigvecs1,
 const lapack_complex_double *eigvecs2,
 const lapack_complex_double *fc3_reciprocal,
 const double *masses,
 const size_t num_atom)
{
  size_t baseIndex, index_l, index_lm, i, j, k, l, m, n;
  double sum_real, sum_imag, mmm, mass_l, mass_lm;
  lapack_complex_double eig_prod, eig_prod1;

  sum_real = 0;
  sum_imag = 0;

  for (l = 0; l < num_atom; l++) {
    mass_l = masses[l];
    index_l = l * num_atom * num_atom * 27;

    for (m = 0; m < num_atom; m++) {
      mass_lm = mass_l * masses[m];
      index_lm = index_l + m * num_atom * 27;

      for (i = 0; i < 3; i++) {
        for (j = 0; j < 3; j++) {
          eig_prod1 = phonoc_complex_prod
            (eigvecs0[(l * 3 + i) * num_atom * 3 + bi0],
             eigvecs1[(m * 3 + j) * num_atom * 3 + bi1]);

          for (n = 0; n < num_atom; n++) {
            mmm = 1.0 / sqrt(mass_lm * masses[n]);
            baseIndex = index_lm + n * 27 + i * 9 + j * 3;

            for (k = 0; k < 3; k++) {
              eig_prod = phonoc_complex_prod
                (eig_prod1, eigvecs2[(n * 3 + k) * num_atom * 3 + bi2]);
              eig_prod = phonoc_complex_prod
                (eig_prod, fc3_reciprocal[baseIndex + k]);
              sum_real += lapack_complex_double_real(eig_prod) * mmm;
              sum_imag += lapack_complex_double_imag(eig_prod) * mmm;
            }
          }
        }
      }
    }
  }
  return lapack_make_complex_double(sum_real, sum_imag);
}

File: phono3py/phonon3/reciprocal_to_normal.py
"""Transform fc3 in reciprocal space to phonon space."""

# Copyright (C) 2020 Atsushi Togo
# All rights reserved.
#
# This file is part of phono3py.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
# * Redistributions of source code must retain the above copyright
#   notice, this list of conditions and the following disclaimer.
#
# * Redistributions in binary form must reproduce the above copyright
#   notice, this list of conditions and the following disclaimer in
#   the documentation and/or other materials provided with the
#   distribution.
#
# * Neither the name of the phonopy project nor the names of its
#   contributors may be used to endorse or promote products derived
#   from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

from __future__ import annotations

from collections.abc import Sequence

import numpy as np
from numpy.typing import NDArray
from phonopy.structure.atoms import PhonopyAtoms


def run_reciprocal_to_normal_squared_rust(
    fc3_reciprocal: NDArray[np.cdouble],
    frequencies: NDArray[np.double],
    eigenvectors: NDArray[np.cdouble],
    masses: NDArray[np.double],
    band_indices: NDArray[np.int64],
    g_pos: NDArray[np.int64],
    cutoff_frequency: float,
    n_out: int,
) -> NDArray[np.double]:
    """Compute ``|fc3_normal|^2 / (f0*f1*f2)`` at a triplet using the Rust backend.

    ``fc3_reciprocal`` uses the atom-first layout
    ``(num_patom, num_patom, num_patom, 3, 3, 3)`` produced by
    ``run_real_to_reciprocal_rust``.  ``frequencies`` has shape
    ``(3, num_band)`` and ``eigenvectors`` has shape
    ``(3, num_band, num_band)``; each row selects the quantity at one
    vertex of the triplet.  ``eigenvectors`` is indexed as
    ``[component, band]`` (un-scaled; mass scaling is applied inside).

    ``g_pos`` has shape ``(num_g_pos, 4)`` with columns
    ``(i0, j, k, dest)`` where ``i0`` indexes ``band_indices`` and
    ``dest`` is the flat offset in the output.  Output length is
    ``n_out``; entries not touched by ``g_pos`` are zero.

    """
    import phono3py_rs

    fc3_normal_squared = np.zeros(n_out, dtype="double")
    f0, f1, f2 = np.ascontiguousarray(frequencies, dtype="double")
    e0, e1, e2 = (np.ascontiguousarray(e, dtype="complex128") for e in eigenvectors)

    phono3py_rs.reciprocal_to_normal_squared(
        fc3_normal_squared,
        np.ascontiguousarray(g_pos, dtype="int64"),
        np.ascontiguousarray(fc3_reciprocal, dtype="complex128"),
        f0,
        f1,
        f2,
        e0,
        e1,
        e2,
        np.ascontiguousarray(masses, dtype="double"),
        np.ascontiguousarray(band_indices, dtype="int64"),
        float(cutoff_frequency),
    )
    return fc3_normal_squared


class ReciprocalToNormal:
    """Class to transform fc3 in reciprocal space to phonon space.

    This is an implementation in python for prototyping and the test.
    Equivalent routine is implemented in C, and this is what usually
    we use.

    """

    def __init__(
        self,
        primitive: PhonopyAtoms,
        frequencies: NDArray[np.double],
        eigenvectors: NDArray[np.cdouble],
        band_indices: Sequence[int] | NDArray[np.int64],
        cutoff_frequency: float = 0,
    ) -> None:
        """Init method."""
        self._primitive = primitive
        self._frequencies = frequencies
        self._eigenvectors = eigenvectors
        self._band_indices = band_indices
        self._cutoff_frequency = cutoff_frequency

        self._fc3_normal: NDArray[np.cdouble]
        self._fc3_reciprocal: NDArray[np.cdouble]

    def run(
        self,
        fc3_reciprocal: NDArray[np.cdouble],
        grid_triplet: NDArray[np.int64],
    ) -> None:
        """Calculate fc3 in phonon coordinates."""
        num_band = len(self._primitive) * 3
        self._fc3_reciprocal = fc3_reciprocal
        dtype = "c%d" % (np.dtype("double").itemsize * 2)
        self._fc3_normal = np.zeros(
            (len(self._band_indices), num_band, num_band), dtype=dtype
        )
        self._reciprocal_to_normal(grid_triplet)

    def get_reciprocal_to_normal(self) -> NDArray[np.cdouble] | None:
        """Return fc3 in phonon coordinates."""
        return self._fc3_normal

    def _reciprocal_to_normal(self, grid_triplet: NDArray[np.int64]) -> None:
        e1, e2, e3 = self._eigenvectors[grid_triplet]
        f1, f2, f3 = self._frequencies[grid_triplet]
        num_band = len(f1)
        cutoff = self._cutoff_frequency
        for i, j, k in list(np.ndindex(len(self._band_indices), num_band, num_band)):
            bi = int(self._band_indices[i])
            if f1[bi] > cutoff and f2[j] > cutoff and f3[k] > cutoff:
                fc3_elem = self._sum_in_atoms((bi, j, k), (e1, e2, e3))
                fff = np.sqrt(f1[bi] * f2[j] * f3[k])
                self._fc3_normal[i, j, k] = fc3_elem / fff

    def _sum_in_atoms(
        self,
        band_indices: tuple[int, int, int],
        eigvecs: tuple[NDArray[np.cdouble], NDArray[np.cdouble], NDArray[np.cdouble]],
    ) -> complex:
        num_atom = len(self._primitive)
        (e1, e2, e3) = eigvecs
        (b1, b2, b3) = band_indices

        sum_fc3 = 0j
        for i, j, k in list(np.ndindex((num_atom,) * 3)):
            sum_fc3_cart = 0
            for ll, m, n in list(np.ndindex((3, 3, 3))):
                sum_fc3_cart += (
                    e1[i * 3 + ll, b1]
                    * e2[j * 3 + m, b2]
                    * e3[k * 3 + n, b3]
                    * self._fc3_reciprocal[i, j, k, ll, m, n]
                )
            mass_sqrt = np.sqrt(np.prod(self._primitive.masses[[i, j, k]]))
            sum_fc3 += sum_fc3_cart / mass_sqrt

        return sum_fc3
